### PR TITLE
Adding Sparse Records support to pycdf

### PR DESCRIPTION
**Pull request by AntoineBrunet**
This PR adds support to the CDF Sparse Records modes in pycdf. It mainly defines the `Var.sparse_records` wrapper to get and set the Sparse Records modes. When a variable is set to any sparse mode, the check for out of range record index is skipped, so it's possible to set data to any record.

This PR also adds a `pad_value` function to get and set the pad value for a given variable, which is important for the PAD_SPARSERECORDS mode.

- [x] Pull request has descriptive title
- [x] Pull request gives overview of changes
- [x] New code has inline comments where necessary
- [x] Any new modules, functions or classes have docstrings consistent with SpacePy style
- [x] Added an entry to CHANGELOG if fixing a major bug or providing a major new feature
- [x] New features and bug fixes should have unit tests
- [x] Relevant issues are linked to (e.g. `See issue #` or `Closes #`)

### CODE REVIEW

**Comment by balarsen** [comment]
Very cool, thanks for the PR! @jtniehof is our CDF expert I will defer to him for any tuning.

**Review by jtniehof — CHANGES_REQUESTED** [review]
Thanks a lot; this is a great enhancement and quite well-executed. I have a lot of really detailed comments because I've tried to be very careful with the pycdf interface. For any or all of these suggestions, I am perfectly happy to implement and push to your branch for further discussion, or leave to you--take your pick!

In addition to the line-based comments, I have a few higher-level ones:

This merits an entry in the release notes, not just the changelog (I'm going to check off the changelog check box, since you did that.)

The CDF `new` method probably should accept a kwarg to set the sparse records mode, like is used for compression.

The new methods need to be tied into the autodocs via the autosummary table at the end of the `Var` docstring. Eventually this will be fully automatic but it isn't right now.

The tests definitely need to be extended. In particular I'm thinking of multi-dimensional variables and insertion/deletion of records. I'm not even sure what *should* happen for insertion and deletion with a sparse variable.

Finally, thinking about the overall API: there are a few different things that change the way a variable "acts." There's the [raw_var](https://spacepy.github.io/autosummary/spacepy.pycdf.CDF.html#spacepy.pycdf.CDF.raw_var) means of accessing a variable which turns off time conversions. That's a pycdf-only thing: it has no effect on the file on disk. But ultimately I do intend to be adding other options like AstroPy time or datetime64. On the other hand, there's the [compress method](https://spacepy.github.io/autosummary/spacepy.pycdf.Var.html#spacepy.pycdf.Var.compress) of variables, which does affect the variable on-disk. Sparse records is very much like `compress` and the API in this PR comes pretty close to that. My concern is that if we keep on adding variable mutators it will get unwieldy and maybe it's time to think about that larger design now. My gut feeling is _probably not_, that it's more reasonable to do a generic interface for conversions (like raw, time type, etc.) but not to try and unite `compress` and `sparse`, and particularly not make them act the same as `raw`. But I wanted to mention it in case it sounded like a good idea to someone else.

Here's the checklist of things to consider, both high-level and in the individual comments. I'm just marking it all out so I don't lose track and we can check off as either commits are made or discussion reveals there's no change to make.

- [ ] Release notes
- [x] Tying in to autosummary
- [ ] Tests for inserting/deleting records and multidimensional
- [ ] Support in `CDF.new`
- [ ] API questions (function to change the sparseness vs. something more generic)
- [ ] Warning check in `Var.__getitem__`
- [ ] Docstring and line spacing for `prepare`
- [ ] Tests for `_prepare`
- [ ] Name `sparse` vs. `sparse_records`
- [ ] More documentation on `sparse_records` method
- [ ] Using CDF error for "wrong kind of sparseness"
- [ ] Document `pad_value` method
- [ ] Test, document, refine dimensional handling on `pad_value`
- [ ] Simplify construction of Hyperslice `__init__` for non-sparse out-of-range; explicitly test

Thanks a lot for putting this in!

File: spacepy/pycdf/__init__.py
Comment (AntoineBrunet) [reply spacepy/pycdf/__init__.py:3115]: My bad, I did not test it properly in non-sparse mode. I see your point about not supressing CDF warnings, though this particular one seems a bit silly to me. I guess this warning could still be useful when reading a file without knowing it contains sparse variable.
I don't mind if we revert this change.

File: spacepy/pycdf/__init__.py
Comment (AntoineBrunet) [reply spacepy/pycdf/__init__.py:3184]: Along this line, there is an assymetry between the input and output conversions, as one is done in the Var class and the other in the Hyperslice class. Overall I found it difficult to put and get values to the CDF side of things, but I had never worked with ctypes before.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3115]: My goal of always passing along warnings from C is a bit funny because the Python and C conventions on warnings/exceptions are different...the C library interface will happily let you ignore warnings (and errors) unless you explicitly handle them, whereas in Python if you don't explicitly handle them they'll make noise. Given that a warning can be suppressed, but a warning never issued can't be retrieved, I would like to go ahead and revert this.

What would make sense (and I'll add it to the punch list) would be to document the warnings handling and how to suppress it. I use a custom warning type specifically to make that easy.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3184]: Honestly, I'm really impressed with this implementation going pretty deep into the guts of pycdf. The input/output asymmetry is partially historical (I think it came about from before pycdf used numpy!) and I'll look into refactoring some of that in the future.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3115]: I made the warning documentation a separate issue (#509) rather than lump it into here.

File: spacepy/pycdf/__init__.py
Comment (AntoineBrunet) [reply spacepy/pycdf/__init__.py:3516]: As far as I can tell the pad value is always a single value, regardless of the variable dimensions. My implementation of pad_value is a bit silly, I was just looking for a way to use the data conversions in Hyperslice.

**Comment by AntoineBrunet** [comment]
I can't reproduce the errors found by the CI (even with Python 3.9 and numpy 1.19.2). I don't know why some versions raise more warnings than other. If you have an idea I'd be glad to hear it. I have corrected some of the issues you have raised.

- [ ] Release notes
- [x] Tying in to autosummary
- [ ] Tests for inserting/deleting records and multidimensional
- [ ] Support in CDF.new
- [ ] API questions (function to change the sparseness vs. something more generic)
- [x] Warning check in Var.__getitem__
- [ ] Docstring and line spacing for prepare
- [ ] Tests for _prepare
- [x] Name sparse vs. sparse_records
- [x] More documentation on sparse_records method
- [x] Using CDF error for "wrong kind of sparseness"
- [x] Document pad_value method
- [ ] Test, document, refine dimensional handling on pad_value
- [ ] Simplify construction of Hyperslice __init__ for non-sparse out-of-range; explicitly test

**Comment by jtniehof** [comment]
Thanks for the quick turnaround. I won't have a chance to dig in today but hopefully tomorrow (US time). A lot of the remaining checkoffs are easy for me to do, if you don't mind me pushing to your PR.

**Comment by AntoineBrunet** [comment]
Please do. Do I need to give you some kind of access to my branch?

Le mer. 17 févr. 2021 à 22:28, Jon Niehof <notifications@github.com> a
écrit :

**Comment by jtniehof** [comment]
I've got more to go, but the changes look good. I made a couple more small docs tweaks and added the release notes. I also marked the API question as done...it doesn't seem like this is the right time.

I did a rebase against master and then force-pushed (you had enabled maintainer edits, which is what's necessary for this.) That would happen at the end of the process anyway, and it gave me access to the new testing infrastructure in #506.

The CI failures were on numpy 1.20, which was starting to raise a bunch more deprecation warnings. The check for specific number of warnings thus failed. I switched to the new `assertWarns` methods so it's not sensitive to exact warning count. I then also went and fixed the warnings. Sort of hijacking your PR for something unrelated but since it was cropping up here, figured I'd fix it here.

I'll come back to the punchlist later but was able to get a few of these done, figured I'd update.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py]: Ah, now I understand it...it's to be able to set up the buffer for the return value. Clever. I was just a little thrown by putting `value` in a list.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py]: Ah, excellent, existing tests catch this case.

**Comment by jtniehof** [comment]
I've gone down the punchlist and looks like everything's done. @AntoineBrunet, please look it over and make sure it's still suitable for your application. The only major change in behavior is I enabled read-off-the-end if sparse records are enabled; it seemed reasonable. Also did some testing and cleanup around NRV variables and unwritten pads. Non-devs can't do a formal "review" but if you comment with a thumbs-up (or thumbs-down) it'd be appreciated.

At this point I've got my fingers all over it so I'm going to ask for a review from another dev. @balarsen, @drsteve, @spacecataz, any takers?

**Comment by AntoineBrunet** [comment]
I have pulled the latest commit, and it works fine for me (I only use the PREV_SPARSERECORDS mode in my usecase). Thanks again for this work, I think it will be very useful.

**Review by jtniehof — APPROVED** [review]
Thanks @AntoineBrunet ! I'm going to approve the PR but give some time for other devs before merging.

**Review by spacecataz — COMMENTED** [review]
Hey, all.

The changes look good; I just have a few notes on some of the docstrings where I think we can be more clear. However, I am not a CDF poweruser, so it very well may be that my comments are not helpful.  

Let me know and then I am happy to approve/merge.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3530]: Yes, should be. I'll make another commit with this change (but later, based on your feedback below.)

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3571]: Agreed.

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3580]: This is a bit sticky, because it's possible not to have a pad value, and then there's no previous one to retain. "Won't be set" maybe?

File: spacepy/pycdf/__init__.py
Comment (jtniehof) [reply spacepy/pycdf/__init__.py:3992]: Both of these are methods of `VarCopy`, which is a full copy of a variable. `VarCopy` doesn't have the concept of sparseness or padding (it's a `dmarray`), so it's just holding what was in the variable from which the `VarCopy` was made. I agree it's confusing but don't have better wording--with that explanation can you think of something?

File: spacepy/pycdf/__init__.py
Comment (spacecataz) [reply spacepy/pycdf/__init__.py:3580]: Yes, that sounds good.  Perhaps, "will not be set/changed" to cover all the scenarios.

File: spacepy/pycdf/__init__.py
Comment (spacecataz) [reply spacepy/pycdf/__init__.py:3992]: Yes, that's helps.  With that context, the original text isn't too far off.  But perhaps a small simplification:
"Gets the pad value of the copied variable"

**Review by jtniehof — APPROVED** [review]
I'm good here. @spacecataz , feel free to approve and merge if this is good for you. And thanks.

**Pull request merged**
